### PR TITLE
WIP: Experimental new value interface

### PR DESCRIPTION
**Pull request by apelisse**
This is not meant to work yet, this is work-in-progress to test having an interface for value so potentially we wouldn't have to convert anything in anything.

### CODE REVIEW

File: value/value.go [hunk 1]
```diff
@@ -182,32 +127,325 @@ func (f FieldList) Less(rhs FieldList) bool {
...
+	Float() float64
+	String() string
+
+	Interface() interface{}
 }
 
-// Less compares two lists lexically.
-func (l *List) Less(rhs *List) bool {
+type List interface {
+	Interface() []interface{}
```
Comment (lavalamp) [inline value/value.go]: I wouldn't have expected this method?

Comment (apelisse) [reply value/value.go]: I don't know if I use it?

File: value/value.go [hunk 2]
```diff
@@ -182,32 +127,325 @@ func (f FieldList) Less(rhs FieldList) bool {
...
+		return true
+	}
+	return false
+}
+
+func (v ValueInterface) Map() Map {
+	if v.Value == nil {
+		return MapString(nil)
+	}
+	switch t := v.Value.(type) {
```
Comment (lavalamp) [inline value/value.go]: I would not expect this?

File: value/value.go [hunk 3]
```diff
@@ -182,32 +127,325 @@ func (f FieldList) Less(rhs FieldList) bool {
...
-// Less compares two lists lexically.
-func (l *List) Less(rhs *List) bool {
+type List interface {
+	Interface() []interface{}
+	Length() int
+	Iterate(func(int, Value))
+	At(int) Value
+}
+
+type ValueInterface struct {
```
Comment (lavalamp) [inline value/value.go]: This is an implementation, not an interface?

File: value/value.go (hunk 2 above)
Comment (jennybuckley) [reply value/value.go]: Didn't this already exist?

Comment (lavalamp) [reply value/value.go]: ah, ok, I think I understand now.

Comment (jennybuckley) [reply value/value.go]: Oh I was only looking at the last commit

File: value/value.go (hunk 3 above)
Comment (lavalamp) [reply value/value.go]: Specifically, this is the k8s unstructured implementation. I think this implementation should go in a different package and get a different name :)

File: value/value.go [hunk 4]
```diff
@@ -319,96 +502,68 @@ func (m *Map) Less(rhs *Map) bool {
...
+		vk, ok := key.(string)
+		if !ok {
+			continue
 		}
-		return "[" + strings.Join(strs, ",") + "]"
-	case v.MapValue != nil:
-		strs := []string{}
-		for _, i := range v.MapValue.Items {
-			strs = append(strs, fmt.Sprintf("%v=%v", i.Name, i.Value))
+		if !iter(vk, ValueInterface{Value: value}) {
```
Comment (lavalamp) [inline value/value.go]: This is the thing that escapes?

Comment (apelisse) [reply value/value.go]: yes

Comment (apelisse) [reply value/value.go]:
```
value/value.go:565:30: ValueInterface literal escapes to heap
value/value.go:565:30:  from iter(vk, ValueInterface literal) (parameter to indirect call) at value/value.go:565:11
```

Comment (lavalamp) [reply value/value.go]: Here's an idea, probably worth what you paid for it:

```
type mapValue struct {
  m map[string]interface{}
  k string
}

...

if !iter(vk, mapValue{m, vk}) {
...
```

Comment (apelisse) [reply value/value.go]: but the mapValue is still going to escape?

Comment (lavalamp) [reply value/value.go]: mapValue is not a pointer though...

Comment (apelisse) [reply value/value.go]: `ValueInterface{}` is not a pointer either?

Comment (lavalamp) [reply value/value.go]: oh wait, it gets moved to the heap when stuffed in an interface? :(

I guess that makes sense.

Comment (apelisse) [reply value/value.go]: Yeah ...

Comment (lavalamp) [reply value/value.go]: If that's the problem, we can allocate a single one and use it for everything?

Comment (apelisse) [reply value/value.go]: I would have to think of the consequences of doing so? Are we sure nothing is going to keep a copy of it somewhere?

Comment (lavalamp) [reply value/value.go]: Hopefully that would break a test...

Comment (apelisse) [reply value/value.go]: I'm pretty sure it would, we can try this out

**Comment by apelisse** [comment]
Replaced by #129

**Pull request closed without merging**